### PR TITLE
feat(integration): add support for sage intacct xml

### PR DESCRIPTION
**Pull request by hassan254-prog**
## Describe your changes

- Add support for sage intacct xml

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

**Comment by hassan254-prog** [comment]
Thanks to @TBonnin, @bodinsamuel and @khaliqgant , we can now comfortably add in support for such, no matter how complex the input body is without needing to add in a new `authmode` 😄

**Comment by hassan254-prog** [comment]
@bodinsamuel, I would appreciate some clarification on how you intend to implement the `TWO_STEP` authentication in connect-ui, whether we should add `credentials` for this in `providers.yaml`. This will help me add the necessary validation for the dynamic credential fields, ensuring that the validation tests pass above.

File: packages/shared/lib/services/connection.service.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import jwt from 'jsonwebtoken';
+import { Builder, parseStringPromise } from 'xml2js';
```
Comment (bodinsamuel) [inline packages/shared/lib/services/connection.service.ts]: can you check other options for the package? this one seems badly maintained, maybe those 2 can do the same:
- https://www.npmjs.com/package/fast-xml-parser (this one has large sponsor)
- https://www.npmjs.com/package/@xmldom/xmldom

File: packages/shared/providers.yaml [hunk 2]
```diff
@@ -4893,6 +4893,65 @@ sage:
...
+    proxy:
+        base_url: https://api.intacct.com/ia/xml/xmlgw.phtml
+    token_url: https://api.intacct.com/ia/xml/xmlgw.phtml
+    body_format: xml
+    token_params:
+        request:
+            control:
+                senderid: ${credential.senderId}
+                password: ${credential.senderPassword}
+                controlid: ${timestamp}
```
Comment (bodinsamuel) [inline packages/shared/providers.yaml]: is it really a timestamp? do you have associated doc?

File: packages/shared/providers.yaml [hunk 3]
```diff
@@ -4893,6 +4893,65 @@ sage:
...
+    docs: https://docs.nango.dev/integrations/all/sage
+    credentials:
+        senderId:
+            type: string
+            title: Sender ID
+            description: Your Sage Intacct Sender ID
+        senderPassword:
+            type: string
+            title: Sender Password
+            description: Your Sage Intacct Sender Password
```
Comment (bodinsamuel) [inline packages/shared/providers.yaml:4981]: secret: true

File: packages/shared/providers.yaml [hunk 4]
```diff
@@ -4893,6 +4893,65 @@ sage:
...
+            title: User ID
+            description: Your Sage Intacct User ID
+        companyId:
+            type: string
+            title: Company ID
+            description: Your Sage Intacct Company ID
+        userPassword:
+            type: string
+            title: User Password
+            description: Your Sage Intacct User Password
```
Comment (bodinsamuel) [inline packages/shared/providers.yaml:4994]: secret: true

File: packages/shared/providers.yaml [hunk 5]
```diff
@@ -4893,6 +4893,65 @@ sage:
...
+            description: Your Sage Intacct Sender Password
+        userId:
+            type: string
+            title: User ID
+            description: Your Sage Intacct User ID
+        companyId:
+            type: string
+            title: Company ID
+            description: Your Sage Intacct Company ID
+        userPassword:
```
Comment (bodinsamuel) [inline packages/shared/providers.yaml:4991]: what's the diff between userPassword and senderPassword?

File: packages/shared/providers.yaml (hunk 3 above)
Comment (hassan254-prog) [reply packages/shared/providers.yaml:4981]: done.

File: packages/shared/providers.yaml (hunk 4 above)
Comment (hassan254-prog) [reply packages/shared/providers.yaml:4994]: updated in latest commit.

File: packages/shared/lib/services/connection.service.ts (hunk 1 above)
Comment (hassan254-prog) [reply packages/shared/lib/services/connection.service.ts]: updated with the latest commit 👍.

File: packages/shared/providers.yaml (hunk 2 above)
Comment (hassan254-prog) [reply packages/shared/providers.yaml]: They recommend using a `timestamp` for this value as it is a good way to differentiate similar [requests](https://developer.intacct.com/web-services/requests/#controlid-element-required). 
https://developer.intacct.com/web-services/your-first-api-calls/

File: packages/shared/providers.yaml (hunk 5 above)
Comment (hassan254-prog) [reply packages/shared/providers.yaml:4991]: The sender password is for the web service developer license, which has a combination of the sender Id and sender password, while the userPassword is part of the login credentials for the company to which you will be sending API calls. 
https://developer.intacct.com/web-services/your-first-api-calls/#prerequisites

File: packages/shared/lib/services/connection.service.ts [hunk 6]
```diff
@@ -1776,15 +1779,36 @@ class ConnectionService {
...
+                bodyFormat === 'xml'
+                    ? new XMLBuilder({
+                          format: true,
+                          indentBy: '  ',
+                          attributeNamePrefix: '$',
+                          ignoreAttributes: false
+                      }).build(postBody)
+                    : JSON.stringify(postBody);
+
+            console.log(bodyContent);
```
Comment (TBonnin) [inline packages/shared/lib/services/connection.service.ts]: 🪓

File: packages/shared/providers.yaml (hunk 2 above)
Comment (bodinsamuel) [reply packages/shared/providers.yaml]: why not 😅

Comment (TBonnin) [reply packages/shared/providers.yaml]: what about `${timestamp.now}` or `${current_timestamp}` or just `${now}` to be clear about what the timestamp value  is going to be

File: packages/shared/providers.yaml (hunk 5 above)
Comment (bodinsamuel) [reply packages/shared/providers.yaml:4991]: Alright, this will be clearer with the doc later

**Review by TBonnin — APPROVED** [review]

**Comment by TBonnin** [comment]
https://github.com/NangoHQ/nango/pull/3014#discussion_r1848450281

File: packages/shared/providers.yaml (hunk 2 above)
Comment (hassan254-prog) [reply packages/shared/providers.yaml]: I have updated this with the latest commit.

File: packages/shared/lib/services/connection.service.ts (hunk 6 above)
Comment (hassan254-prog) [reply packages/shared/lib/services/connection.service.ts]: I have updated this with the new commit 🤦‍♂️, thank you.

**Comment by hassan254-prog** [comment]
@bodinsamuel , I have made some small addition with the latest [commit](https://github.com/NangoHQ/nango/pull/3014/commits/93fd8dc0e519772fa3939d2f3239fde0c9f9b921), to now include credentials to `TWO_STEP`, auth mode, that way we wont have to painfully back fill them later.

**Review by bodinsamuel — APPROVED** [review]

**Comment by bodinsamuel** [comment]
> that way we wont have to painfully back fill them later.

Nobody reads the logs, if you want it to be enforced you can add `error = true`

**Comment by hassan254-prog** [comment]
> > that way we wont have to painfully back fill them later.
> 
> Nobody reads the logs, if you want it to be enforced you can add `error = true`

Thank you Samuel 🙏.

**Pull request merged**